Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -369,6 +377,30 @@ inline bool ShouldDraw() {
         return CurrentGState().ShouldDraw();
     }
 
+    inline void PushBeginDraw() {
+        if ((_beginEndDrawDepth)++ == 0) {
+            deviceContext->BeginDraw();
+        }
+    }
+
+    inline void PopEndDraw() {
+        if (--(_beginEndDrawDepth) == 0) {
+            THROW_IF_FAILED(deviceContext->EndDraw());
+        }
+    }
+
+    inline void EscapeBeginEndDrawStack() {
+        if ((_beginEndDrawDepth > 0) && ((_escapeBeginEndDrawDepth)++ == 0)) {
+            THROW_IF_FAILED(deviceContext->EndDraw());
```

Comment:
The standardization we agreed on was to capture exceptions at the API boundary (which is usually the API entry point).  Failing fast in the leaf nodes will lead to code that just arbitrarily quits with no possible recovery.  Particularly when drawing to hardware device contexts, EndDraw() can fail, so we need to handle this with care.  See #1194.